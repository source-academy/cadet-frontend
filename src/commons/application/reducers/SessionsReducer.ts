import { Reducer } from 'redux';

import { SourceActionType } from '../../utils/ActionsHelper';
import { defaultSession } from '../ApplicationTypes';
import {
  SET_GAME_STATE,
  UPDATE_MATERIAL_DIRECTORY_TREE,
  UPDATE_MATERIAL_INDEX
} from '../types/ActionTypes';
import { LOG_OUT } from '../types/CommonsTypes';
import {
  SessionState,
  SET_TOKENS,
  SET_USER,
  UPDATE_ASSESSMENT,
  UPDATE_ASSESSMENT_OVERVIEWS,
  UPDATE_GRADING,
  UPDATE_GRADING_OVERVIEWS,
  UPDATE_HISTORY_HELPERS,
  UPDATE_NOTIFICATIONS
} from '../types/SessionTypes';

export const SessionsReducer: Reducer<SessionState> = (
  state = defaultSession,
  action: SourceActionType
) => {
  switch (action.type) {
    case LOG_OUT:
      return defaultSession;
    case SET_TOKENS:
      return {
        ...state,
        accessToken: action.payload.accessToken,
        refreshToken: action.payload.refreshToken
      };
    case SET_USER:
      return {
        ...state,
        ...action.payload
      };
    case UPDATE_HISTORY_HELPERS:
      const helper = state.historyHelper;
      const isAcademy = isAcademyRe.exec(action.payload) != null;
      const newAcademyLocations = isAcademy
        ? [helper.lastAcademyLocations[1], action.payload]
        : [...helper.lastAcademyLocations];
      const newGeneralLocations = [helper.lastGeneralLocations[1], action.payload];
      return {
        ...state,
        historyHelper: {
          lastAcademyLocations: newAcademyLocations,
          lastGeneralLocations: newGeneralLocations
        }
      };
    case UPDATE_ASSESSMENT:
      const newAssessments = new Map(state.assessments);
      newAssessments.set(action.payload.id, action.payload);
      return {
        ...state,
        assessments: newAssessments
      };
    case UPDATE_ASSESSMENT_OVERVIEWS:
      return {
        ...state,
        assessmentOverviews: action.payload
      };
    case UPDATE_GRADING:
      const newGradings = new Map(state.gradings);
      newGradings.set(action.payload.submissionId, action.payload.grading);
      return {
        ...state,
        gradings: newGradings
      };
    case UPDATE_GRADING_OVERVIEWS:
      return {
        ...state,
        gradingOverviews: action.payload
      };
    case UPDATE_MATERIAL_DIRECTORY_TREE:
      return {
        ...state,
        materialDirectoryTree: action.payload.directoryTree
      };
    case UPDATE_MATERIAL_INDEX:
      return {
        ...state,
        materialIndex: action.payload.index
      };
    case UPDATE_NOTIFICATIONS:
      return {
        ...state,
        notifications: action.payload
      };
    case SET_GAME_STATE:
      return {
        ...state,
        gameState: action.payload
      };
    default:
      return state;
  }
};

export const isAcademyRe = new RegExp('^/academy.*', 'i');
