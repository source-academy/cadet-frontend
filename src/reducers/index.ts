import { reducer as academy } from './academy';
import { reducer as application } from './application';
import { reducer as playground } from './playground';
import { reducer as session } from './session';
import { reducer as workspaces } from './workspaces';

export default {
  academy,
  application,
  playground,
  session,
  workspaces
};
