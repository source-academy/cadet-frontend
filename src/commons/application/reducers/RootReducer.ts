import { connectRouter } from 'connected-react-router';
import { History } from 'history';
import { combineReducers } from 'redux';

import { AcademyReducer as academy } from '../../../features/academy/AcademyReducer';
import { DashboardReducer as dashboard } from '../../../features/dashboard/DashboardReducer';
import { PlaygroundReducer as playground } from '../../../features/playground/PlaygroundReducer';
import { WorkspaceReducer as workspaces } from '../../workspace/WorkspaceReducer';
import { ApplicationReducer as application } from '../ApplicationReducer';
import { SessionsReducer as session } from './SessionsReducer';

const createRootReducer = (history: History) =>
  combineReducers({
    router: connectRouter(history),
    academy,
    application,
    dashboard,
    playground,
    session,
    workspaces
  });

export default createRootReducer;
