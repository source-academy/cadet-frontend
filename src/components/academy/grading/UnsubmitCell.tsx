import { Alert, Intent } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import * as React from 'react';

import { Role } from '../../../reducers/states';
import { controlButton } from '../../commons';
import { GradingOverview } from './gradingShape';

export interface IUnsubmitCellProps {
  data: GradingOverview;
  handleUnsubmitSubmission: (submissionId: number) => void;
  group: string | null;
  role?: Role;
}

type UnsubmitCellState = {
  isAlertOpen: boolean;
};

class UnsubmitCell extends React.Component<IUnsubmitCellProps, UnsubmitCellState> {
  public constructor(props: IUnsubmitCellProps) {
    super(props);

    this.state = {
      isAlertOpen: false
    };
  }

  public render() {
    if (this.props.data.submissionStatus !== 'submitted') {
      return null;
    } else if (this.props.group !== this.props.data.groupName && this.props.role! !== Role.Admin) {
      return null;
    }

    return (
      <div>
        <Alert
          canEscapeKeyCancel={true}
          canOutsideClickCancel={true}
          cancelButtonText="Cancel"
          className="unsubmit-alert alert"
          intent={Intent.DANGER}
          onConfirm={this.handleConfirmUnsubmit}
          isOpen={this.state.isAlertOpen}
          onCancel={this.handleUnsubmitAlertClose}
        >
          Are you sure you want to unsubmit?
        </Alert>
        {controlButton('', IconNames.ARROW_LEFT, () => this.setState({ isAlertOpen: true }), {
          fullWidth: true
        })}
      </div>
    );
  }

  private handleConfirmUnsubmit = () => {
    this.props.handleUnsubmitSubmission(this.props.data.submissionId);
    this.setState({ isAlertOpen: false });
  };

  private handleUnsubmitAlertClose = () => this.setState({ isAlertOpen: false });
}

export default UnsubmitCell;
