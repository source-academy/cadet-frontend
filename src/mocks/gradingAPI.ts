import { Grading, GradingOverview } from '../components/academy/grading/gradingShape';
import { mock2DRuneLibrary as mockLibrary } from './assessmentAPI';
import { mockFetchRole, Role, Roles } from './userAPI';

export const mockGradingOverviews: GradingOverview[] = [
  {
    gradeAdjustment: 0,
    xpAdjustment: 0,
    assessmentCategory: 'Mission',
    assessmentId: 0,
    assessmentName: 'Mission 0 ',
    currentGrade: 69,
    currentXp: 69,
    xpBonus: 10,
    initialGrade: 69,
    initialXp: 69,
    maxGrade: 100,
    maxXp: 100,
    studentId: 0,
    studentName: 'Al Gorithm',
    submissionId: 0,
    groupName: '1D'
  },
  {
    gradeAdjustment: -2,
    xpAdjustment: -2,
    assessmentCategory: 'Mission',
    assessmentId: 1,
    assessmentName: 'Mission 1',
    currentGrade: -2,
    currentXp: -2,
    xpBonus: 12,
    initialGrade: 0,
    initialXp: 0,
    maxGrade: 400,
    maxXp: 400,
    studentId: 0,
    studentName: 'Dee Sign',
    submissionId: 1,
    groupName: '1F'
  },
  {
    gradeAdjustment: 4,
    xpAdjustment: 4,
    assessmentCategory: 'Mission',
    assessmentId: 0,
    assessmentName: 'Mission 0',
    currentGrade: 1000,
    currentXp: 1000,
    xpBonus: 12,
    initialGrade: 996,
    initialXp: 996,
    maxGrade: 1000,
    maxXp: 1000,
    studentId: 1,
    studentName: 'May Trix',
    submissionId: 2,
    groupName: '1F'
  }
];

/**
 * Mock for fetching a trainer/admin's student grading information.
 * A null value is returned for invalid token or role.
 *
 * @param accessToken a valid access token for the cadet backend.
 * @param group a boolean if true, only fetches submissions from the grader's group
 */
export const mockFetchGradingOverview = (
  accessToken: string,
  group: boolean
): GradingOverview[] | null => {
  // mocks backend role fetching
  const permittedRoles: Role[] = [Roles.admin, Roles.trainer];
  const role: Role | null = mockFetchRole(accessToken);
  if (role === null || !permittedRoles.includes(role)) {
    return null;
  } else {
    return group ? [mockGradingOverviews[0]] : mockGradingOverviews;
  }
};

const mockGrading: Grading = [
  {
    question: {
      answer: "This student's answer to the 0th question......",
      content: `
Hello and welcome to this assessment! This is the *0th question*.

\`\`\`
>>> import this
\`\`\`
`,
      comment: null,
      id: 0,
      library: mockLibrary,
      solutionTemplate: '0th question mock solution template',
      solution: 'This is how the 0th question is `solved`',
      type: 'programming',
      maxGrade: 1000,
      maxXp: 1000,
      grader: {
        name: 'avenger',
        id: 1
      },
      gradedAt: '2038-06-18T05:24:26.026Z',
      xp: 1,
      grade: 1
    },
    grade: {
      gradeAdjustment: 0,
      xpAdjustment: 0,
      grade: 0,
      xp: 0,
      comment: 'Good job!!'
    },
    student: {
      name: 'Al Gorithm',
      id: 0
    }
  },
  {
    question: {
      answer: "This student's answer to the 1st question",
      comment: null,
      content: 'Hello and welcome to this assessment! This is the 1st question.',
      id: 1,
      library: mockLibrary,
      solutionTemplate: '1st question mock solution template',
      solution: null,
      type: 'programming',
      maxGrade: 200,
      maxXp: 200,
      grader: {
        name: 'avenger',
        id: 1
      },
      gradedAt: '2038-06-18T05:24:26.026Z',
      xp: 1,
      grade: 1
    },
    grade: {
      gradeAdjustment: 0,
      xpAdjustment: 0,
      grade: 100,
      xp: 100,
      comment: 'Good job!!'
    },
    student: {
      name: 'Al Gorithm',
      id: 0
    }
  },
  {
    question: {
      // C is the answer
      answer: 3,
      comment: null,
      solution: 2,
      content:
        'Hello and welcome to this assessment! This is the 2nd question. Oddly enough, it is an MCQ question!',
      choices: [
        {
          content: 'A',
          hint: 'hint A'
        },
        {
          content: 'B',
          hint: 'hint B'
        },
        {
          content: 'C',
          hint: 'hint C'
        },
        {
          content: 'D',
          hint: 'hint D'
        }
      ],
      id: 2,
      library: mockLibrary,
      type: 'mcq',
      maxGrade: 100,
      maxXp: 100,
      grader: {
        name: 'avenger',
        id: 1
      },
      gradedAt: '2038-06-18T05:24:26.026Z',
      xp: 1,
      grade: 1
    },
    grade: {
      gradeAdjustment: 0,
      xpAdjustment: 0,
      grade: 50,
      xp: 50,
      comment:
        'A Very long string. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer a leo et lectus gravida sagittis a non neque. Phasellus consectetur arcu vitae metus vulputate commodo. Phasellus varius sollicitudin quam a porta. Pellentesque mollis molestie felis vitae imperdiet. Nam porta purus ac tellus luctus ultrices. Integer pellentesque nisl vel nunc ullamcorper, in vehicula est dapibus. Nunc dapibus neque dolor, ut mattis massa mattis in. Fusce nec risus nec ex pharetra lacinia. Mauris sit amet ullamcorper sapien. Suspendisse scelerisque neque sed nunc tincidunt, ac semper enim efficitur. Ut sit amet eleifend arcu. Donec viverra at justo vitae eleifend. Morbi ut erat ultricies, hendrerit mi ut, ornare mauris.'
    },
    student: {
      name: 'Al Gorithm',
      id: 0
    }
  }
];

/**
 * Mock for fetching a trainer/admin's student grading information.
 * A null value is returned for invalid token or role.
 *
 * @param accessToken a valid access token for the cadet backend.
 */
export const mockFetchGrading = (accessToken: string, submissionId: number): Grading | null => {
  // mocks backend role fetching
  const permittedRoles: Role[] = [Roles.admin, Roles.trainer];
  const role: Role | null = mockFetchRole(accessToken);
  if (role === null || !permittedRoles.includes(role)) {
    return null;
  } else {
    return mockGrading;
  }
};
