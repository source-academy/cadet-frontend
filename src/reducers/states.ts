import { Context } from 'js-slang'
import { SourceError } from 'js-slang/dist/types'

import { WorkspaceLocation, WorkspaceLocations } from '../actions/workspaces'
import { Grading, GradingOverview } from '../components/academy/grading/gradingShape'
import { Announcement } from '../components/Announcements'
import {
  ExternalLibraryName,
  ExternalLibraryNames,
  IAssessment,
  IAssessmentOverview
} from '../components/assessment/assessmentShape'
import { HistoryHelper } from '../utils/history'
import { createContext } from '../utils/slangHelper'

export interface IState {
  readonly academy: IAcademyState
  readonly application: IApplicationState
  readonly playground: IPlaygroundState
  readonly session: ISessionState
  readonly workspaces: IWorkspaceManagerState
}

export interface IAcademyState {
  readonly gameCanvas?: HTMLCanvasElement
}

export interface IApplicationState {
  readonly title: string
  readonly environment: ApplicationEnvironment
}

export interface IPlaygroundState {
  readonly queryString?: string
}

interface IAssessmentWorkspace extends IWorkspaceState {
  readonly currentAssessment?: number
  readonly currentQuestion?: number
  readonly hasUnsavedChanges: boolean
}

interface IGradingWorkspace extends IWorkspaceState {
  readonly currentSubmission?: number
  readonly currentQuestion?: number
  readonly hasUnsavedChanges: boolean
}

export interface IPlaygroundWorkspace extends IWorkspaceState {
  readonly playgroundExternal: ExternalLibraryName
}

export interface IWorkspaceManagerState {
  readonly assessment: IAssessmentWorkspace
  readonly grading: IGradingWorkspace
  readonly playground: IPlaygroundWorkspace
}

export interface IWorkspaceState {
  readonly context: Context
  readonly editorValue: string | null
  readonly editorWidth: string
  readonly isEditorAutorun: boolean
  readonly isRunning: boolean
  readonly output: InterpreterOutput[]
  readonly replHistory: ReplHistory
  readonly replValue: string
  readonly sideContentActiveTab: number
  readonly sideContentHeight?: number
  readonly globals: Array<[string, any]>
}

export interface ISessionState {
  readonly accessToken?: string
  readonly assessmentOverviews?: IAssessmentOverview[]
  readonly assessments: Map<number, IAssessment>
  readonly announcements?: Announcement[]
  readonly grade: number
  readonly gradingOverviews?: GradingOverview[]
  readonly gradings: Map<number, Grading>
  readonly historyHelper: HistoryHelper
  readonly maxGrade: number
  readonly maxXp: number
  readonly refreshToken?: string
  readonly role?: Role
  readonly story?: Story
  readonly name?: string
  readonly xp: number
}

type ReplHistory = {
  browseIndex: null | number // [0, 49] if browsing, else null
  records: string[]
}

export const maxBrowseIndex = 50

export type Story = {
  story: string
  playStory: boolean
}

/**
 * An output while the program is still being run in the interpreter. As a
 * result, there are no return values or SourceErrors yet. However, there could
 * have been calls to display (console.log) that need to be printed out.
 */
export type RunningOutput = {
  type: 'running'
  consoleLogs: string[]
}

/**
 * An output which reflects the program which the user had entered. Not a true
 * Output from the interpreter, but simply there to let he user know what had
 * been entered.
 */
export type CodeOutput = {
  type: 'code'
  value: string
}

/**
 * An output which represents a program being run successfully, i.e. with a
 * return value at the end. A program can have either a return value, or errors,
 * but not both.
 */
export type ResultOutput = {
  type: 'result'
  value: any
  consoleLogs: string[]
  runtime?: number
  isProgram?: boolean
}

/**
 * An output which represents a program being run unsuccessfully, i.e. with
 * errors at the end. A program can have either a return value, or errors, but
 * not both.
 */
export type ErrorOutput = {
  type: 'errors'
  errors: SourceError[]
  consoleLogs: string[]
}

export type InterpreterOutput = RunningOutput | CodeOutput | ResultOutput | ErrorOutput

export enum ApplicationEnvironment {
  Development = 'development',
  Production = 'production',
  Test = 'test'
}

export enum Role {
  Student = 'student',
  Staff = 'staff',
  Admin = 'admin'
}

/**
 * Defines what chapters are available for usage.
 * For external libraries, see externalLibraries.ts
 */
export const sourceChapters = [1, 2, 3, 4]
const latestSourceChapter = sourceChapters.slice(-1)[0]

const currentEnvironment = (): ApplicationEnvironment => {
  switch (process.env.NODE_ENV) {
    case 'development':
      return ApplicationEnvironment.Development
    case 'production':
      return ApplicationEnvironment.Production
    default:
      return ApplicationEnvironment.Test
  }
}

export const defaultAcademy: IAcademyState = {
  gameCanvas: undefined
}

export const defaultApplication: IApplicationState = {
  title: 'Cadet',
  environment: currentEnvironment()
}

export const defaultPlayground: IPlaygroundState = {}

export const defaultEditorValue = '// Type your program in here!'

/**
 * Create a default IWorkspaceState for 'resetting' a workspace.
 * Takes in parameters to set the js-slang library and chapter.
 *
 * @param location the location of the workspace, used for context
 */
export const createDefaultWorkspace = (location: WorkspaceLocation): IWorkspaceState => ({
  context: createContext<WorkspaceLocation>(latestSourceChapter, [], location),
  editorValue: location === WorkspaceLocations.playground ? defaultEditorValue : null,
  editorWidth: '50%',
  output: [],
  replHistory: {
    browseIndex: null,
    records: []
  },
  replValue: '',
  sideContentActiveTab: 0,
  globals: [],
  isEditorAutorun: false,
  isRunning: false
})

export const defaultComments = 'Comments **here**. Use `markdown` if you ~~are cool~~ want!'

export const defaultWorkspaceManager: IWorkspaceManagerState = {
  assessment: {
    ...createDefaultWorkspace(WorkspaceLocations.assessment),
    currentAssessment: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  grading: {
    ...createDefaultWorkspace(WorkspaceLocations.grading),
    currentSubmission: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  playground: {
    ...createDefaultWorkspace(WorkspaceLocations.playground),
    playgroundExternal: ExternalLibraryNames.NONE
  }
}

export const defaultSession: ISessionState = {
  accessToken: undefined,
  announcements: [
    {
      author: 'Aministrator',
      title: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
      content:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam non tellus non ligula consectetur feugiat a at mauris. Ut sagittis, urna id suscipit dictum, ipsum diam sollicitudin enim, sed ultricies diam turpis vel libero. Maecenas finibus nisl ac lobortis ultricies. Integer molestie urna vitae nisi pharetra porttitor. Vestibulum massa diam, tristique quis ante a, posuere placerat magna. Pellentesque at lobortis purus, vitae imperdiet diam. Nulla eu rutrum neque. Aliquam efficitur consectetur ullamcorper. Pellentesque ultricies, diam ut vestibulum pellentesque, metus arcu laoreet ex, at mattis quam est non neque. Nam velit ipsum, posuere non porttitor commodo, lobortis in urna. Nulla facilisi. Donec mollis id nibh a luctus. Mauris vitae orci id velit pulvinar pellentesque non ut sapien. Curabitur eu consequat lorem. Proin pretium blandit scelerisque. Morbi ultricies, tellus non posuere pretium, felis sem convallis magna, ut sagittis elit felis et sem. Aliquam auctor suscipit condimentum. Nam posuere nulla dolor, in maximus risus feugiat vel. Phasellus vestibulum odio nec leo vehicula condimentum. Ut et semper libero. Cras fermentum mauris quis ex sodales, sit amet hendrerit augue lobortis. Maecenas eu dapibus enim, nec auctor est. Quisque quis erat erat. Curabitur sed rutrum felis, non venenatis diam. Fusce maximus rhoncus neque, in maximus velit semper eu. Fusce tempus lorem ut sodales pulvinar.',
      pinned: true
    }
  ],
  assessments: new Map<number, IAssessment>(),
  assessmentOverviews: undefined,
  grade: 0,
  gradingOverviews: undefined,
  gradings: new Map<number, Grading>(),
  historyHelper: {
    lastAcademyLocations: [null, null],
    lastGeneralLocations: [null, null]
  },
  maxGrade: 0,
  maxXp: 0,
  refreshToken: undefined,
  name: undefined,
  xp: 0
}

export const defaultState: IState = {
  academy: defaultAcademy,
  application: defaultApplication,
  playground: defaultPlayground,
  session: defaultSession,
  workspaces: defaultWorkspaceManager
}
