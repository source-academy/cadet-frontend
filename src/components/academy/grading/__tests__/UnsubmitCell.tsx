import { mount } from 'enzyme';
import * as React from 'react';

import { mockGradingOverviews as gradingOverviews } from '../../../../mocks/gradingAPI';
import { Role } from '../../../../reducers/states';
import UnsubmitCell, { IUnsubmitCellProps } from '../UnsubmitCell';

// Pre-condition: only Staff and Admin users can view the grading overviews page

// Default props has: submitted submission with group 1D, no user group and Staff user
const defaultUnsubmitCellProps: IUnsubmitCellProps = {
  data: gradingOverviews[2],
  handleUnsubmitSubmission: (submissionId: number) => {},
  group: null,
  role: Role.Staff
};

test('Unsubmit cell does not render for unsubmitted submissions', () => {
  const props = {
    ...defaultUnsubmitCellProps,
    data: { ...gradingOverviews[2], submissionStatus: 'attempted' },
    role: Role.Admin
  };
  const app = <UnsubmitCell {...props} />;
  const tree = mount(app);
  expect(tree.debug()).toMatchSnapshot();
  expect(tree.find('*').hostNodes().length).toEqual(0);
});

test("Unsubmit cell does not render for submitted submissions outside of non-Admin user's group", () => {
  const props = {
    ...defaultUnsubmitCellProps,
    group: '1D'
  };
  const app = <UnsubmitCell {...props} />;
  const tree = mount(app);
  expect(tree.debug()).toMatchSnapshot();
  expect(tree.find('*').hostNodes().length).toEqual(0);
});

test('Unsubmit cell renders for submitted submissions for Admin users regardless of group', () => {
  const props = {
    ...defaultUnsubmitCellProps,
    role: Role.Admin
  };
  const app = <UnsubmitCell {...props} />;
  const tree = mount(app);
  expect(tree.debug()).toMatchSnapshot();
  expect(tree.find('*').hostNodes().length).toBeGreaterThan(0);
  expect(tree.find('.bp3-button').hostNodes()).toHaveLength(1);
  expect(
    tree
      .find('.unsubmit-alert')
      .at(0)
      .prop('isOpen')
  ).toEqual(false);
});

test("Unsubmit cell renders for submitted submissions belonging to a Staff user's group", () => {
  const props = {
    ...defaultUnsubmitCellProps,
    group: '1F'
  };
  const app = <UnsubmitCell {...props} />;
  const tree = mount(app);
  expect(tree.debug()).toMatchSnapshot();
  expect(tree.find('*').hostNodes().length).toBeGreaterThan(0);
  expect(tree.find('.bp3-button').hostNodes()).toHaveLength(1);
  expect(
    tree
      .find('.unsubmit-alert')
      .at(0)
      .prop('isOpen')
  ).toEqual(false);
});
