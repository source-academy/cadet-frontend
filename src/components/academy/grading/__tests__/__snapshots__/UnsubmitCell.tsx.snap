// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unsubmit cell does not render for submitted submissions outside of non-Admin user's group 1`] = `"<UnsubmitCell data={{...}} handleUnsubmitSubmission={[Function: handleUnsubmitSubmission]} group=\\"1D\\" role=\\"staff\\" />"`;

exports[`Unsubmit cell does not render for unsubmitted submissions 1`] = `"<UnsubmitCell data={{...}} handleUnsubmitSubmission={[Function: handleUnsubmitSubmission]} group={{...}} role=\\"admin\\" />"`;

exports[`Unsubmit cell renders for submitted submissions belonging to a Staff user's group 1`] = `
"<UnsubmitCell data={{...}} handleUnsubmitSubmission={[Function: handleUnsubmitSubmission]} group=\\"1F\\" role=\\"staff\\">
  <div>
    <Blueprint3.Alert canEscapeKeyCancel={true} canOutsideClickCancel={true} cancelButtonText=\\"Cancel\\" className=\\"unsubmit-alert alert\\" intent=\\"danger\\" onConfirm={[Function]} isOpen={false} onCancel={[Function]} confirmButtonText=\\"OK\\">
      <Blueprint3.Dialog onConfirm={[Function]} isOpen={false} onCancel={[Function]} className=\\"bp3-alert unsubmit-alert alert\\" canEscapeKeyClose={true} canOutsideClickClose={true} onClose={[Function]} portalContainer={[undefined]}>
        <Blueprint3.Overlay onConfirm={[Function]} isOpen={false} onCancel={[Function]} className=\\"bp3-overlay-scroll-container\\" canEscapeKeyClose={true} canOutsideClickClose={true} onClose={[Function]} portalContainer={[undefined]} hasBackdrop={true} autoFocus={true} backdropProps={{...}} enforceFocus={true} lazy={true} transitionDuration={300} transitionName=\\"bp3-overlay\\" usePortal={true} />
      </Blueprint3.Dialog>
    </Blueprint3.Alert>
    <Blueprint3.Button disabled={false} fill={true} intent=\\"none\\" minimal={true} className=\\"\\" type={[undefined]} icon={{...}} onClick={[Function]}>
      <button type={[undefined]} disabled={[undefined]} className=\\"bp3-button bp3-fill bp3-minimal\\" onClick={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} tabIndex={[undefined]}>
        <Blueprint3.Icon icon={{...}}>
          <Blueprint3.Icon icon=\\"arrow-left\\" color={[undefined]}>
            <span icon=\\"arrow-left\\" className=\\"bp3-icon bp3-icon-arrow-left\\" title={[undefined]}>
              <svg fill={[undefined]} data-icon=\\"arrow-left\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
                <desc>
                  arrow-left
                </desc>
                <path d=\\"M13.99 6.99H4.41L7.7 3.7a1.003 1.003 0 0 0-1.42-1.42l-5 5a1.014 1.014 0 0 0 0 1.42l5 5a1.003 1.003 0 0 0 1.42-1.42L4.41 8.99H14c.55 0 1-.45 1-1s-.46-1-1.01-1z\\" fillRule=\\"evenodd\\" />
              </svg>
            </span>
          </Blueprint3.Icon>
        </Blueprint3.Icon>
        <Blueprint3.Icon icon={[undefined]} />
      </button>
    </Blueprint3.Button>
  </div>
</UnsubmitCell>"
`;

exports[`Unsubmit cell renders for submitted submissions for Admin users regardless of group 1`] = `
"<UnsubmitCell data={{...}} handleUnsubmitSubmission={[Function: handleUnsubmitSubmission]} group={{...}} role=\\"admin\\">
  <div>
    <Blueprint3.Alert canEscapeKeyCancel={true} canOutsideClickCancel={true} cancelButtonText=\\"Cancel\\" className=\\"unsubmit-alert alert\\" intent=\\"danger\\" onConfirm={[Function]} isOpen={false} onCancel={[Function]} confirmButtonText=\\"OK\\">
      <Blueprint3.Dialog onConfirm={[Function]} isOpen={false} onCancel={[Function]} className=\\"bp3-alert unsubmit-alert alert\\" canEscapeKeyClose={true} canOutsideClickClose={true} onClose={[Function]} portalContainer={[undefined]}>
        <Blueprint3.Overlay onConfirm={[Function]} isOpen={false} onCancel={[Function]} className=\\"bp3-overlay-scroll-container\\" canEscapeKeyClose={true} canOutsideClickClose={true} onClose={[Function]} portalContainer={[undefined]} hasBackdrop={true} autoFocus={true} backdropProps={{...}} enforceFocus={true} lazy={true} transitionDuration={300} transitionName=\\"bp3-overlay\\" usePortal={true} />
      </Blueprint3.Dialog>
    </Blueprint3.Alert>
    <Blueprint3.Button disabled={false} fill={true} intent=\\"none\\" minimal={true} className=\\"\\" type={[undefined]} icon={{...}} onClick={[Function]}>
      <button type={[undefined]} disabled={[undefined]} className=\\"bp3-button bp3-fill bp3-minimal\\" onClick={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} tabIndex={[undefined]}>
        <Blueprint3.Icon icon={{...}}>
          <Blueprint3.Icon icon=\\"arrow-left\\" color={[undefined]}>
            <span icon=\\"arrow-left\\" className=\\"bp3-icon bp3-icon-arrow-left\\" title={[undefined]}>
              <svg fill={[undefined]} data-icon=\\"arrow-left\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
                <desc>
                  arrow-left
                </desc>
                <path d=\\"M13.99 6.99H4.41L7.7 3.7a1.003 1.003 0 0 0-1.42-1.42l-5 5a1.014 1.014 0 0 0 0 1.42l5 5a1.003 1.003 0 0 0 1.42-1.42L4.41 8.99H14c.55 0 1-.45 1-1s-.46-1-1.01-1z\\" fillRule=\\"evenodd\\" />
              </svg>
            </span>
          </Blueprint3.Icon>
        </Blueprint3.Icon>
        <Blueprint3.Icon icon={[undefined]} />
      </button>
    </Blueprint3.Button>
  </div>
</UnsubmitCell>"
`;
